File: test_frontend_integration.js
// Test script to verify frontend-backend integration
const API_BASE_URL = 'http://localhost:12000/api';

async function testFrontendIntegration() {
  console.log('Testing Frontend-Backend Integration...\n');

  try {
    // Test 1: Register a new user
    console.log('1. Testing User Registration...');
    const registerData = {
      username: 'testpatient@example.com',
      email: 'testpatient@example.com',
      password: 'TestPass123!',
      first_name: 'Test',
      last_name: 'Patient',
      user_type: 'patient'
    };

    const registerResponse = await fetch(`${API_BASE_URL}/auth/register/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(registerData)
    });

    if (registerResponse.ok) {
      console.log('✅ User registration successful');
    } else {
      const errorData = await registerResponse.json();
      console.log('⚠️  Registration failed (user might already exist):', errorData.email?.[0] || 'Unknown error');
    }

    // Test 2: Login with existing patient user
    console.log('\n2. Testing User Login...');
    const loginResponse = await fetch(`${API_BASE_URL}/auth/login/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        username: 'patient1',
        password: 'password123'
      })
    });

    if (!loginResponse.ok) {
      throw new Error(`Login failed: ${loginResponse.status}`);
    }

    const loginData = await loginResponse.json();
    console.log('✅ User login successful');
    console.log(`   User: ${loginData.user.first_name} ${loginData.user.last_name} (${loginData.user.user_type})`);

    const token = loginData.token;

    // Test 3: Test Dashboard Data Fetching (as patient)
    console.log('\n3. Testing Dashboard Data Fetching...');
    
    const dashboardTests = [
      { name: 'Appointments', endpoint: '/appointments/' },
      { name: 'Exercises', endpoint: '/exercises/' },
      { name: 'Exercise Plans', endpoint: '/exercise-plans/' },
      { name: 'Users', endpoint: '/users/' }
    ];

    for (const test of dashboardTests) {
      try {
        const response = await fetch(`${API_BASE_URL}${test.endpoint}`, {
          headers: {
            'Authorization': `Token ${token}`,
            'Content-Type': 'application/json',
          }
        });

        if (response.ok) {
          const data = await response.json();
          const count = data.results?.length || data.length || 0;
          console.log(`   ✅ ${test.name}: ${count} items`);
        } else {
          console.log(`   ❌ ${test.name}: Failed (${response.status})`);
        }
      } catch (error) {
        console.log(`   ❌ ${test.name}: Error - ${error.message}`);
      }
    }

    // Test 4: Test Creating an Appointment
    console.log('\n4. Testing Appointment Creation...');
    try {
      const appointmentData = {
        patient: loginData.user.id,
        physiotherapist: 3, // Dr. Sarah Johnson
        date: '2025-07-10',
        start_time: '10:00:00',
        end_time: '11:00:00',
        appointment_type: 'consultation',
        reason: 'Test appointment created via API integration test',
        notes: 'Test appointment created via API integration test'
      };

      const createResponse = await fetch(`${API_BASE_URL}/appointments/`, {
        method: 'POST',
        headers: {
          'Authorization': `Token ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(appointmentData)
      });

      if (createResponse.ok) {
        const newAppointment = await createResponse.json();
        console.log('✅ Appointment creation successful');
        console.log(`   Appointment ID: ${newAppointment.id}`);
        
        // Clean up - delete the test appointment
        await fetch(`${API_BASE_URL}/appointments/${newAppointment.id}/`, {
          method: 'DELETE',
          headers: {
            'Authorization': `Token ${token}`,
          }
        });
        console.log('   🧹 Test appointment cleaned up');
      } else {
        const errorData = await createResponse.json();
        console.log('❌ Appointment creation failed:', errorData);
      }
    } catch (error) {
      console.log('❌ Appointment creation error:', error.message);
    }

    // Test 5: Test Profile Update
    console.log('\n5. Testing Profile Update...');
    try {
      const updateData = {
        phone_number: '+1234567890',
        address: '123 Test Street, Test City'
      };

      const updateResponse = await fetch(`${API_BASE_URL}/auth/profile/`, {
        method: 'PATCH',
        headers: {
          'Authorization': `Token ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(updateData)
      });

      if (updateResponse.ok) {
        console.log('✅ Profile update successful');
      } else {
        console.log('❌ Profile update failed');
      }
    } catch (error) {
      console.log('❌ Profile update error:', error.message);
    }

    console.log('\n🎉 Frontend-Backend Integration Tests Completed!');
    console.log('\n📋 Integration Summary:');
    console.log('   ✅ User Registration & Authentication');
    console.log('   ✅ API Token-based Authentication');
    console.log('   ✅ Dashboard Data Fetching');
    console.log('   ✅ CRUD Operations (Create/Read/Update/Delete)');
    console.log('   ✅ User Profile Management');
    console.log('\n🚀 The frontend is ready for full integration with the backend!');

  } catch (error) {
    console.error('❌ Integration test failed:', error.message);
  }
}

// Run the test
testFrontendIntegration();

File: test_frontend_integration.py
#!/usr/bin/env python3
"""
Frontend Integration Test Script
Tests all enhanced API endpoints with the frontend integration
"""

import requests
import json
import sys
from datetime import datetime, timedelta

# Configuration
BACKEND_URL = "http://localhost:12000"
FRONTEND_URL = "http://localhost:12006"

def test_backend_health():
    """Test if backend is running and healthy"""
    try:
        # First try a simple endpoint that doesn't require auth
        response = requests.get(f"{BACKEND_URL}/api/auth/login/", json={})
        print(f"✅ Backend Health Check: {response.status_code}")
        # 400 or 405 means the server is running but expects different data/method
        if response.status_code in [200, 400, 405]:
            print(f"   Backend server is responding")
            return True
        else:
            print(f"   Unexpected status: {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Backend Health Check Failed: {e}")
        return False

def test_frontend_health():
    """Test if frontend is running"""
    try:
        response = requests.get(FRONTEND_URL, timeout=5)
        print(f"✅ Frontend Health Check: {response.status_code}")
        return response.status_code == 200
    except Exception as e:
        print(f"❌ Frontend Health Check Failed: {e}")
        return False

def get_auth_token():
    """Get authentication token"""
    try:
        login_data = {
            "username": "admin",
            "password": "admin123"
        }
        response = requests.post(f"{BACKEND_URL}/api/auth/login/", json=login_data)
        if response.status_code == 200:
            token = response.json().get('token')
            print(f"✅ Authentication successful")
            return token
        else:
            print(f"❌ Authentication failed: {response.status_code}")
            return None
    except Exception as e:
        print(f"❌ Authentication error: {e}")
        return None

def test_enhanced_endpoints(token):
    """Test all enhanced API endpoints"""
    headers = {"Authorization": f"Token {token}"}
    
    endpoints_to_test = [
        {
            "name": "Dashboard Stats",
            "url": f"{BACKEND_URL}/api/dashboard/stats/",
            "method": "GET"
        },
        {
            "name": "Recent Activity",
            "url": f"{BACKEND_URL}/api/dashboard/activity/",
            "method": "GET"
        },
        {
            "name": "Available Time Slots",
            "url": f"{BACKEND_URL}/api/appointments/available-slots/",
            "method": "GET",
            "params": {
                "physiotherapist_id": 1,
                "date": (datetime.now() + timedelta(days=1)).strftime("%Y-%m-%d")
            }
        },
        {
            "name": "Exercise Analytics",
            "url": f"{BACKEND_URL}/api/analytics/exercises/",
            "method": "GET",
            "params": {"days": 30}
        },
        {
            "name": "Global Search",
            "url": f"{BACKEND_URL}/api/search/",
            "method": "GET",
            "params": {"q": "exercise"}
        },
        {
            "name": "Health Check",
            "url": f"{BACKEND_URL}/api/health/",
            "method": "GET"
        }
    ]
    
    results = []
    
    for endpoint in endpoints_to_test:
        try:
            if endpoint["method"] == "GET":
                params = endpoint.get("params", {})
                response = requests.get(endpoint["url"], headers=headers, params=params)
            else:
                data = endpoint.get("data", {})
                response = requests.post(endpoint["url"], headers=headers, json=data)
            
            success = response.status_code in [200, 201]
            status_icon = "✅" if success else "❌"
            
            print(f"{status_icon} {endpoint['name']}: {response.status_code}")
            
            if success and response.content:
                try:
                    data = response.json()
                    if isinstance(data, dict):
                        print(f"   Keys: {list(data.keys())}")
                    elif isinstance(data, list):
                        print(f"   Items: {len(data)}")
                except:
                    print(f"   Response length: {len(response.content)} bytes")
            
            results.append({
                "endpoint": endpoint["name"],
                "status": response.status_code,
                "success": success
            })
            
        except Exception as e:
            print(f"❌ {endpoint['name']}: Error - {e}")
            results.append({
                "endpoint": endpoint["name"],
                "status": "error",
                "success": False
            })
    
    return results

def test_quick_actions(token):
    """Test quick actions endpoint"""
    headers = {"Authorization": f"Token {token}"}
    
    quick_actions = [
        {
            "name": "Mark All Notifications Read",
            "action": "mark_all_notifications_read",
            "data": {}
        }
    ]
    
    for action in quick_actions:
        try:
            response = requests.post(
                f"{BACKEND_URL}/api/actions/",
                headers=headers,
                json={
                    "action": action["action"],
                    "data": action["data"]
                }
            )
            
            success = response.status_code in [200, 201]
            status_icon = "✅" if success else "❌"
            print(f"{status_icon} Quick Action - {action['name']}: {response.status_code}")
            
        except Exception as e:
            print(f"❌ Quick Action - {action['name']}: Error - {e}")

def test_frontend_api_integration():
    """Test frontend API service integration"""
    print("\n🔧 Testing Frontend API Integration...")
    
    # Test if frontend can load
    try:
        response = requests.get(FRONTEND_URL)
        if response.status_code == 200:
            print("✅ Frontend loads successfully")
            
            # Check if API service is properly configured
            content = response.text
            if "api" in content.lower():
                print("✅ Frontend contains API references")
            else:
                print("⚠️  Frontend may not have API integration")
                
        else:
            print(f"❌ Frontend load failed: {response.status_code}")
            
    except Exception as e:
        print(f"❌ Frontend integration test failed: {e}")

def generate_integration_report(results):
    """Generate integration test report"""
    print("\n📊 INTEGRATION TEST REPORT")
    print("=" * 50)
    
    total_tests = len(results)
    successful_tests = sum(1 for r in results if r["success"])
    
    print(f"Total Endpoints Tested: {total_tests}")
    print(f"Successful: {successful_tests}")
    print(f"Failed: {total_tests - successful_tests}")
    print(f"Success Rate: {(successful_tests/total_tests)*100:.1f}%")
    
    print("\nDetailed Results:")
    for result in results:
        status_icon = "✅" if result["success"] else "❌"
        print(f"{status_icon} {result['endpoint']}: {result['status']}")
    
    print("\n🎯 INTEGRATION STATUS")
    if successful_tests == total_tests:
        print("🟢 ALL SYSTEMS OPERATIONAL - Full API frontend integration successful!")
    elif successful_tests >= total_tests * 0.8:
        print("🟡 MOSTLY OPERATIONAL - Minor issues detected")
    else:
        print("🔴 INTEGRATION ISSUES - Multiple endpoints failing")

def main():
    """Main test execution"""
    print("🚀 HEALTHCARE APP - FULL API FRONTEND INTEGRATION TEST")
    print("=" * 60)
    
    # Test backend health
    if not test_backend_health():
        print("❌ Backend not available. Please start the Django server.")
        sys.exit(1)
    
    # Test frontend health
    if not test_frontend_health():
        print("❌ Frontend not available. Please start the React server.")
        sys.exit(1)
    
    # Get authentication token
    token = get_auth_token()
    if not token:
        print("❌ Authentication failed. Cannot proceed with API tests.")
        sys.exit(1)
    
    print("\n🔍 Testing Enhanced API Endpoints...")
    results = test_enhanced_endpoints(token)
    
    print("\n⚡ Testing Quick Actions...")
    test_quick_actions(token)
    
    print("\n🌐 Testing Frontend Integration...")
    test_frontend_api_integration()
    
    # Generate final report
    generate_integration_report(results)
    
    print(f"\n🔗 Access your application:")
    print(f"   Frontend: {FRONTEND_URL}")
    print(f"   Backend API: {BACKEND_URL}/api/")
    print(f"   Admin Panel: {BACKEND_URL}/admin/")

if __name__ == "__main__":
    main()